src/main/java/learnositysdk/request/DataApi.java

```diff
@@ -209,21 +198,41 @@ public void requestRecursive(RequestCallback callback) throws Exception
 
 			if (this.responseHasNext(body)) {
 				this.requestPacket.put("next", meta.get("next"));
-				this.requestString = this.requestPacket.toString();
+				this.updateRequestString();
 				makeNextRequest = true;
 			}
 		} while (makeNextRequest);
 	}
 
-    private JSONObject createResponseObject(Remote remote) throws Exception
-    {
-    	JSONObject response = new JSONObject();
-    	response.put("body", remote.getBody());
-    	response.put("contentType",  remote.getContentType());
-    	response.put("statusCode", remote.getStatusCode());
-    	response.put("timeTaken", remote.getTimeTaken());
-    	return response;
-    }
+	/**
+	 * Sets the request packet, updating the stored string version as well
+	 * @param requestPacket The request packet to store
+	 */
+	private void setRequestPacket(JSONObject requestPacket)
+	{
+		if (Telemetry.isEnabled()) {
+			Telemetry.addToRequest(requestPacket);
```

Comment:
Could we return the updated request packet, rather than modifying it in place?
```java
this.requestPacket = Telemetry.addToRequest(requestPacket);
```

Then no need to do it just below (actually, no need to do anything down below, I guess the requestString only needs update when the requestPacket has been updated.

Reply:
The request string will still need to be updated - especially upon first call. But since the check is done in the Telemetry class we can definitely update this to an assignment without the `if` around it.

Reply:
I find the code very confusing (maybe it always was). Can't you just let Init do all the work on construction (i.e. add telemetry, create the request string and generate the signature)? Then in functions request and requestRecursive, you could just call Init.getRequestString. That way there is no need for duplicated code in Data. Also there wouldn't be any need to check in Init if altering the request packet is allowed or not.

Reply:
So my suggestion would be something like this: here https://github.com/Learnosity/learnosity-sdk-java/blob/LRN-20223/feature/add-telemetry/src/main/java/learnositysdk/request/DataApi.java#L116 just set the request packet (don't set the string). You might want to validate if you want to throw the exception early, but validation is redone when constructing Init anyway. Then in function request, you check if your requestObject is null (no longer a check on the string). If not you initialise Init with the requestObject. That will validate and add telemetry. Then when putting the request to the options, you call Init.getRequestString() and that should match the generated signature.

Reply:
RESOLVED: We've cleaned this up a little bit with the new approach in the polish commit.